libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -837,6 +779,20 @@ bool TypeInference::visit(TypeDefinition const& _typeDefinition)
 	else
 		typeDefinitionAnnotation.type = helper.typeFunctionType(helper.tupleType(arguments), definedType);
 
+	std::optional<Type> underlyingType;
+
+	if (_typeDefinition.typeExpression())
+	{
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		_typeDefinition.typeExpression()->accept(*this);
+
+		if (!isBuiltIn)
```

Comment:
If you move this ``if`` outside you can only visit the ``typeExpression()`` if ``isBuiltIn`` is false, which would make the bogus ``visit(BuiltinDefinition const&)`` redundant and would give us the safety of rejecting builtins in unexpected context in ``visitNode``...